core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveEmptyExceptBranches.java

```diff
@@ -0,0 +1,94 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.ExceptNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+
+import java.util.List;
+
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isEmpty;
+import static io.trino.sql.planner.plan.Patterns.except;
+
+/**
+ * Removes branches from an ExceptNode that are guaranteed to produce 0 rows.
+ *
+ * <code>A EXCEPT E1 EXCEPT E2 ... EXCEPT En</code> is equivalent to <code>A - (E1 + ... + En)</code>.
+ * If any of the Ei sets is empty, it can be removed. If only A is left, the Except operation gets
+ * replaced with a projection to preserve the outputs symbols.
+ */
+public class RemoveEmptyExceptBranches
+        implements Rule<ExceptNode>
+{
+    private static final Pattern<ExceptNode> PATTERN = except();
+
+    @Override
+    public Pattern<ExceptNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(ExceptNode node, Captures captures, Context context)
+    {
+        boolean hasEmptyBranches = node.getSources().stream()
+                .skip(1) // first source is the set we're excluding rows from, so ignore it
+                .anyMatch(source -> isEmpty(source, context.getLookup()));
+
+        if (!hasEmptyBranches) {
+            return Result.empty();
+        }
+
+        ImmutableList.Builder<PlanNode> newSourcesBuilder = ImmutableList.builder();
+        ImmutableListMultimap.Builder<Symbol, Symbol> outputsToInputsBuilder = ImmutableListMultimap.builder();
+
+        for (int i = 0; i < node.getSources().size(); i++) {
+            PlanNode source = node.getSources().get(i);
+            if (i == 0 || !isEmpty(source, context.getLookup())) {
+                newSourcesBuilder.add(source);
+
+                for (Symbol column : node.getOutputSymbols()) {
+                    outputsToInputsBuilder.put(column, node.getSymbolMapping().get(column).get(i));
+                }
+            }
+        }
+
+        List<PlanNode> newSources = newSourcesBuilder.build();
+        ListMultimap<Symbol, Symbol> outputsToInputs = outputsToInputsBuilder.build();
+
+        if (newSources.size() == 1) {
+            Assignments.Builder assignments = Assignments.builder();
+
+            outputsToInputs.entries().stream()
+                    .forEach(entry -> assignments.put(entry.getKey(), entry.getValue().toSymbolReference()));
+
+            return Result.ofPlanNode(
+                    new ProjectNode(
```

Comment:
If the original `ExceptNode` was DISTINCT, the result should be distinct:
`{1, 1, 1, 2} EXCEPT DISTINCT {} --> {1, 2}`
It requires adding an aggregation.

Reply:
Good catch!